### PR TITLE
MueLu: RefMaxwell diverse

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/muelu 

## Description
- Forward hierarchy label to Tpetra in RAPFactory: Important when several MueLu hierarchies are built
- Fix performance bug in Thyra-RefMaxwell factory
- Make shifted Laplacian experimental (Issue #2608)
- Fix build error in RefMaxwell when Epetra in on, but no matching Tpetra instantiation is available (Issue #3032)

### CODE REVIEW

File: packages/muelu/src/Misc/MueLu_ThresholdAFilterFactory_def.hpp [hunk 1]
```diff
@@ -81,8 +81,9 @@ namespace MueLu {
     RCP<OMatrix> Ain = Get< RCP<OMatrix> >(currentLevel, varName_);
 
     // create new empty Matrix
-    RCP<CrsOMatrix> Aout = rcp(new CrsOMatrix(Ain->getRowMap(),Ain->getGlobalMaxNumRowEntries(),Xpetra::StaticProfile)); //FIXME
-
+    RCP<CrsOMatrix> Aout = rcp(new CrsOMatrix(Ain->getRowMap(),Ain->getGlobalMaxNumRowEntries() ? expectedNNZperRow_ <= 0 : expectedNNZperRow_ ,Xpetra::DynamicProfile)); //FIXME
```
Comment (jhux2) [inline packages/muelu/src/Misc/MueLu_ThresholdAFilterFactory_def.hpp]: This looks backwards to me.  Isn't the condition being tested `expectedNNZperRow_ <= 0`?

File: packages/muelu/src/Misc/MueLu_ThresholdAFilterFactory_def.hpp [hunk 2]
```diff
@@ -81,8 +81,9 @@ namespace MueLu {
...
+    Teuchos::RCP<const Map> rowmap = Ain->getRowMap();
```
Comment (jhux2) [inline packages/muelu/src/Misc/MueLu_ThresholdAFilterFactory_def.hpp]: Make this a reference.

File: packages/muelu/src/Misc/MueLu_ThresholdAFilterFactory_def.hpp [hunk 3]
```diff
@@ -81,8 +81,9 @@ namespace MueLu {
...
+    Teuchos::RCP<const Map> colmap = Ain->getColMap();
```
Comment (jhux2) [inline packages/muelu/src/Misc/MueLu_ThresholdAFilterFactory_def.hpp]: Make this a reference.

File: packages/muelu/adapters/xpetra/MueLu_RefMaxwell_decl.hpp [hunk 4]
```diff
@@ -76,7 +76,7 @@
 #include "Xpetra_CrsMatrixWrap_fwd.hpp"
 #include "Xpetra_ExportFactory_fwd.hpp"
 
-#ifdef HAVE_MUELU_IFPACK2
+#if defined(HAVE_MUELU_IFPACK2) && (!defined(HAVE_MUELU_EPETRA) || defined(HAVE_MUELU_TPETRA_INST_INT_INT))
```
Comment (jhux2) [inline packages/muelu/adapters/xpetra/MueLu_RefMaxwell_decl.hpp]: I don't understand why the second condition is required.

Comment (jhux2) [reply packages/muelu/adapters/xpetra/MueLu_RefMaxwell_decl.hpp]: I mean the case (If Ifpack2  and not Epetra).

Comment (cgcgcg) [reply packages/muelu/adapters/xpetra/MueLu_RefMaxwell_decl.hpp]: We need Ifpack2 for Hiptmair smoothing. Also either we want no Epetra, in which case the Tpetra::MultiVector below is defined, or, if we have Epetra, then we want the corresponding Tpetra instantiation to be available.

File: packages/muelu/src/Misc/MueLu_ThresholdAFilterFactory_def.hpp (hunk 1 above)
Comment (cgcgcg) [reply packages/muelu/src/Misc/MueLu_ThresholdAFilterFactory_def.hpp]: Oh dear, that's what happens with too much python..

File: packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp [hunk 5]
```diff
@@ -207,11 +213,13 @@ namespace Thyra {
 
         if (paramList.isParameter("M1")) {
           if (paramList.isType<Teuchos::RCP<TpCrsMat> >("M1")) {
```
Comment (jhux2) [inline packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp]: Is it possible that M1 is neither a TpCrsMat or a ThyLinOpBase?

File: packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp [hunk 6]
```diff
@@ -228,11 +236,13 @@ namespace Thyra {
 
         if (paramList.isParameter("D0")) {
           if (paramList.isType<Teuchos::RCP<TpCrsMat> >("D0")) {
```
Comment (jhux2) [inline packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp]: Same question as above.

File: packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp [hunk 7]
```diff
@@ -249,38 +259,21 @@ namespace Thyra {
 
         if (paramList.isParameter("M0inv")) {
           if (paramList.isType<Teuchos::RCP<TpCrsMat> >("M0inv")) {
```
Comment (jhux2) [inline packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp]: Similar question as above.

File: packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp (hunk 5 above)
Comment (cgcgcg) [reply packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp]: In the _decl, we have a specialization for Epetra that checks for Epetra matrix type as well, if that's what you are referring to.

Comment (jhux2) [reply packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp]: @cgcgcg What I meant is that the `if-then` in file `Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp`  line 214 is of the form
```cpp
         if (paramList.isParameter("M1")) {
￼            if (paramList.isType<Teuchos::RCP<TpCrsMat> >("M1")) {
              Teuchos::TimeMonitor tM_M1(*Teuchos::TimeMonitor::getNewTimer(std::string("ThyraMueLuRefMaxwell::initializePrec wrap M1")));
￼              RCP<TpCrsMat> tM1 = paramList.get<RCP<TpCrsMat> >("M1");
￼              paramList.remove("M1");
￼              RCP<XpCrsMat> xM1 = rcp_dynamic_cast<XpCrsMat>(tM1, true);
￼              paramList.set<RCP<XpCrsMat> >("M1", xM1);
￼            } else if (paramList.isType<Teuchos::RCP<const ThyLinOpBase> >("M1")) {
￼              Teuchos::TimeMonitor tM_M1(*Teuchos::TimeMonitor::getNewTimer(std::string("ThyraMueLuRefMaxwell::initializePrec wrap M1")));
￼              RCP<const ThyLinOpBase> thyM1 = paramList.get<RCP<const ThyLinOpBase> >("M1");
￼              paramList.remove("M1");
￼              RCP<const XpCrsMat> crsM1 = XpThyUtils::toXpetra(thyM1);
              TEUCHOS_TEST_FOR_EXCEPT(Teuchos::is_null(crsM1));
              // MueLu needs a non-const object as input
              RCP<XpCrsMat> crsM1NonConst = Teuchos::rcp_const_cast<XpCrsMat>(crsM1);
              TEUCHOS_TEST_FOR_EXCEPT(Teuchos::is_null(crsM1NonConst));
              // wrap as an Xpetra::Matrix that MueLu can work with
              RCP<XpMat> M1 = rcp(new Xpetra::CrsMatrixWrap<Scalar,LocalOrdinal,GlobalOrdinal,Node>(crsM1NonConst));
              paramList.set<RCP<XpMat> >("M1", M1);
           }
```

Could there a case where M1 is a parameter, but neither of the `if` checks is true?

Comment (cgcgcg) [reply packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp]: Ah, yes. We should throw then.

**Comment by cgcgcg** [comment]
@trilinos/framework The autotester on this PR is failing because the intel build cannot get a license: https://testing-vm.sandia.gov/cdash/viewConfigure.php?buildid=3744213

**Comment by cgcgcg** [comment]
Thanks, @jhux2

**Pull request merged**